File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/PaymentStrategy/CreditCardStrategy.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.PaymentStrategy;

/// <summary>
///     A payment strategy for credit card payments.
/// </summary>
public sealed class CreditCardStrategy : PaymentStrategyBase
{
    private readonly string _cardNumber;

    /// <summary>
    ///     Initializes a new instance of the <see cref="CreditCardStrategy" /> class.
    /// </summary>
    /// <param name="cardNumber">The credit card number.</param>
    /// <exception cref="ArgumentException">Thrown when card number is empty or non-numeric.</exception>
    public CreditCardStrategy(string cardNumber)
    {
        if (string.IsNullOrWhiteSpace(cardNumber) || !cardNumber.All(char.IsDigit))
            throw new ArgumentException("Card number must be numeric and at least 1 digit.", nameof(cardNumber));
        _cardNumber = cardNumber;
    }

    /// <inheritdoc />
    public override void Checkout(InventoryItem item, int quantity, TextWriter output)
    {
        var total = ComputeTotalAmount(item, quantity);
        output.WriteLine($"Checking out {total:C} using Credit Card (Number: {_cardNumber}).");
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/payment/CreditCardStrategy.java
package assignment2solution.domain.payment;

import assignment2solution.domain.inventory.InventoryItem;

import java.io.PrintWriter;
import java.io.Writer;
import java.text.NumberFormat;
import java.util.Locale;
import java.util.Objects;

/**
 * A payment strategy for credit card payments.
 */
public final class CreditCardStrategy extends PaymentStrategyBase {
    private final String cardNumber;
    private final NumberFormat currencyFormatter = NumberFormat.getCurrencyInstance(Locale.US);

    /**
     * Initializes a new instance of the {@link CreditCardStrategy} class.
     */
    public CreditCardStrategy(String cardNumber) {
        if (cardNumber == null || cardNumber.trim().isEmpty() || !cardNumber.chars().allMatch(Character::isDigit)) {
            throw new IllegalArgumentException("Card number must be numeric and at least 1 digit.");
        }
        this.cardNumber = cardNumber;
    }

    @Override
    public void checkout(InventoryItem item, int quantity, Writer output) {
        Objects.requireNonNull(item, "item");
        Objects.requireNonNull(output, "output");
        var writer = output instanceof PrintWriter ? (PrintWriter) output : new PrintWriter(output, true);
        var total = computeTotalAmount(item, quantity);
        writer.println("Checking out " + currencyFormatter.format(total) + " using Credit Card (Number: "
            + cardNumber + ").");
    }
}
